x/subscription/keeper/cu_tracker.go

```diff
@@ -5,34 +5,188 @@ import (
+				utils.Attribute{Key: "block", Value: strconv.FormatUint(entryBlock, 10)},
+			)
+		}
+		trackedCuList = append(trackedCuList, trackedCuInfo{
+			provider:  provider,
+			trackedCu: cu,
+			chainID:   chainID,
+			block:     entryBlock,
+		})
+		totalCuUsedBySub += cu
+	}
+
+	return trackedCuList, totalCuUsedBySub
+}
+
+// remove only before the sub is deleted
+func (k Keeper) RewardAndResetCuTracker(ctx sdk.Context, cuTrackerTimerKeyBytes []byte) {
+	sub, shouldRemove := types.DecodeCuTrackerTimerKey(string(cuTrackerTimerKeyBytes))
+	trackedCuList, totalCuUsedBySub := k.getSubTrackedCuInfo(ctx, sub)
+
+	for _, trackedCuInfo := range trackedCuList {
+		trackedCu := trackedCuInfo.trackedCu
+		provider := trackedCuInfo.provider
+		chainID := trackedCuInfo.chainID
+		block := trackedCuInfo.block
+
+		plan, err := k.GetPlanFromSubscription(ctx, sub, block)
+		if err != nil {
+			utils.LavaFormatError("cannot find subscription's plan", types.ErrCuTrackerPayoutFailed,
+				utils.Attribute{Key: "sub_consumer", Value: sub},
+			)
+			return
+		}
+
+		err = k.resetCuTracker(ctx, sub, trackedCuInfo, shouldRemove)
+		if err != nil {
+			utils.LavaFormatError("removing/reseting tracked CU entry failed", err,
+				utils.Attribute{Key: "provider", Value: provider},
+				utils.Attribute{Key: "tracked_cu", Value: trackedCu},
+				utils.Attribute{Key: "chain_id", Value: chainID},
+				utils.Attribute{Key: "sub", Value: sub},
+				utils.Attribute{Key: "block", Value: ctx.BlockHeight()},
+			)
+			return
+		}
+
+		// TODO: deal with the reward's remainder (uint division...)
+		if trackedCu > totalCuUsedBySub {
+			utils.LavaFormatError("tracked CU of provider is larger than total of CU used by subscription", types.ErrCuTrackerPayoutFailed,
+				utils.Attribute{Key: "provider", Value: provider},
+				utils.Attribute{Key: "tracked_cu", Value: trackedCu},
+				utils.Attribute{Key: "chain_id", Value: chainID},
+				utils.Attribute{Key: "sub", Value: sub},
+				utils.Attribute{Key: "sub_total_used_cu", Value: totalCuUsedBySub},
+				utils.Attribute{Key: "block", Value: strconv.FormatUint(uint64(ctx.BlockHeight()), 10)},
+			)
+			return
+		}
+
+		// monthly reward = (tracked_CU / total_CU_used_in_sub_this_month) * plan_price
+		totalMonthlyReward := plan.Price.Amount.MulRaw(int64(trackedCu)).QuoRaw(int64(totalCuUsedBySub))
+
+		// calculate the provider reward (smaller than totalMonthlyReward because it's shared with delegators)
+		providerAddr, err := sdk.AccAddressFromBech32(provider)
+		if err != nil {
+			utils.LavaFormatError("invalid provider address", err,
+				utils.Attribute{Key: "provider", Value: provider},
+			)
+			return
+		}
+
+		// Note: if the reward function doesn't reward the provider because he was unstaked, we only print an error and not returning
+		providerReward, err := k.dualstakingKeeper.RewardProvidersAndDelegators(ctx, providerAddr, chainID, totalMonthlyReward, types.ModuleName)
+		if err == dualstakingtypes.ErrProviderNotStaked {
```

Comment:
this if else should for warning or error should be inside of RewardProvidersAndDelegators

Reply:
but the caller needs to know why the function failed. Do you mean I should add some sort of flag to mark if a provider is not staked?

Reply:
i mean that the underlying method, RewardProvidersAndDelegators,  should return the specific error

Reply:
Following up our offline discussion, not changing it since I need to know if to return or not according to the error